aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -39,22 +39,38 @@ struct sdp_params {
   bool is_causal;
 };
 
+inline bool check_requires_grad(sdp_params params, bool debug) {
+  const bool any_inputs_require_grad = params.query.requires_grad() ||
+      params.key.requires_grad() || params.value.requires_grad();
+  const bool gradmode_enabled = at::GradMode::is_enabled();
+  if ((any_inputs_require_grad && gradmode_enabled)) {
+    if (debug) {
+      TORCH_WARN("Flash Attention does not currently support training.");
+    }
+    return false;
+  }
+  return true;
+}
+
 inline std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
   constexpr std::array<SDPBackend, num_backends> default_order{
       SDPBackend::flash_attention,
       SDPBackend::efficient_attention,
       SDPBackend::math};
+
+  constexpr std::array<SDPBackend, num_backends> efficient_first{
+      SDPBackend::efficient_attention,
+      SDPBackend::flash_attention,
+      SDPBackend::math};
   // Logic is taken from xformers
   // FlashAttention parallelizes across "batch_size * num_heads"
   // MemEff parallelizes across "batch_size * num_heads * num_queries" and can
   // be more efficient. batch_size, q_len, num_heads, k = inp.query.shape
+
   if (params.query.is_nested() || params.key.is_nested() ||
       params.value.is_nested()) {
     // See check_for_nested_inputs for details
-    return {
```

Comment:
this shortcut path can/should be removed with: https://github.com/pytorch/pytorch/pull/95438

However the `query_lengths{params.query.sym_size(2)}` will error with NTs. This seems like this will also be the case for the new heuristic but just a note to self really